crates/ide-completion/src/tests/flyimport.rs

```diff
@@ -374,6 +374,135 @@ fn main() {
     );
 }
 
+#[test]
+fn trait_method_fuzzy_completion_aware_of_fundamental_boxes() {
+    let fixture = r#"
+//- /fundamental.rs crate:fundamental
+#[lang = "owned_box"]
+#[fundamental]
+pub struct Box<T>(T);
+//- /foo.rs crate:foo
+pub trait TestTrait {
+    fn some_method(&self);
+}
+//- /main.rs crate:main deps:foo,fundamental
+struct TestStruct;
+
+impl foo::TestTrait for fundamental::Box<TestStruct> {
+    fn some_method(&self) {}
+}
+
+fn main() {
+    let t = fundamental::Box(TestStruct);
+    t.$0
+}
+"#;
+
+    check(
+        fixture,
+        expect![[r#"
+        me some_method() (use foo::TestTrait) fn(&self)
+    "#]],
+    );
+
+    check_edit(
+        "some_method",
+        fixture,
+        r#"
+use foo::TestTrait;
+
+struct TestStruct;
+
+impl foo::TestTrait for fundamental::Box<TestStruct> {
+    fn some_method(&self) {}
+}
+
+fn main() {
+    let t = fundamental::Box(TestStruct);
+    t.some_method()$0
+}
+"#,
+    );
+}
+
+#[test]
+fn trait_method_fuzzy_completion_aware_of_fundamental_references() {
+    let fixture = r#"
+//- /foo.rs crate:foo
+pub trait TestTrait {
+    fn some_method(&self);
+}
+//- /main.rs crate:main deps:foo
+struct TestStruct;
+
+impl foo::TestTrait for &TestStruct {
+    fn some_method(&self) {}
+}
+
+fn main() {
+    let t = &TestStruct;
```

Comment:
agreed! weirdly enough, that test fails, despite a rust-analyzer binary working in this case. not sure why, i'm digging, but i don't get an answer shortly, i'm inclined to figure that out later.

Reply:
Given that the completion currently works, I don't think we should break it in the name of perf.

Reply:
Might be because you need `//- minicore: deref`, not sure but could be that autoderef exits early without the lang item present

Reply:
`//- minicore: deref` didn't work, unfortunately :/. I'm inclined to treat this variant as *not* a blocker and investigate later. Here's my summary:
- `impl ForeignTrait for LocalStruct` and `impl ForeignTrait for &LocalStruct` provide completions on `let t = &TestStruct` in _both_ tests and a running binary.
- `impl ForeignTrait for &LocalStruct` only provides completions for provide completions on `let t = TestStruct` in a running binary. 
    - This makes me think that there's a bug in `minicore`—or that I might not be understanding some important details—because this behavioral split reproduces with and without my filtering changes.

Reply:
Made an issue: https://github.com/rust-lang/rust-analyzer/issues/16685